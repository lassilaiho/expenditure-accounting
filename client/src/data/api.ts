export type LoadState =
  | 'not-started'
  | 'loading'
  | 'finished'
  | 'failed';

export class SessionToken {
  public constructor(
    public readonly token: string,
    public readonly expiryTime: Date,
  ) { }

  public static fromJson(json: any): SessionToken | null {
    if (typeof json.token !== 'string'
      || typeof json.expiryTime !== 'string') {
      throw new Error('Invalid json object');
    }
    return new SessionToken(json.token, new Date(json.expiryTime));
  }

  public get isValid() {
    return this.expiryTime.getTime() > Date.now();
  }

  public toJSON(): unknown {
    return {
      token: this.token,
      expiryTime: this.expiryTime.toUTCString(),
    };
  }
}

export default class Api {
  public baseUrl: string = '';

  public defaultHeaders: Record<string, string> = {
    'Authorization': ''
  };

  private _sessionToken: SessionToken | null = null;
  public get sessionToken() { return this._sessionToken; }
  public set sessionToken(token: SessionToken | null) {
    this._sessionToken = token;
    this.defaultHeaders['Authorization'] =
      token === null ? '' : 'Basic ' + btoa(token.token);
  }

  public constructor(baseUrl: string) {
    this.baseUrl = baseUrl;
  }

  public async get(url: string) {
    return fetch(this.baseUrl + url, {
      mode: 'cors',
      headers: { ...this.defaultHeaders },
    });
  }

  public async postJson(url: string, body: any) {
    return fetch(this.baseUrl + url, {
      mode: 'cors',
      method: 'POST',
      headers: {
        ...this.defaultHeaders,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    });
  }

  public async patchJson(url: string, body: any) {
    return fetch(this.baseUrl + url, {
      mode: 'cors',
      method: 'PATCH',
      headers: {
        ...this.defaultHeaders,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    });
  }

  public async delete(url: string) {
    return fetch(this.baseUrl + url, {
      mode: 'cors',
      method: 'DELETE',
      headers: { ...this.defaultHeaders },
    });
  }
}
