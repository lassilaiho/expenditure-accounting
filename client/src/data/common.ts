import { createAction } from '@reduxjs/toolkit';

import { Product } from './products';
import { Purchase } from './purchases';
import { Tag } from './tags';

export type Collection<T> = {
  byId: { [id: number]: T };
  all: number[];
};

export function emptyCollection<T>(): Collection<T> {
  return { byId: {}, all: [] };
}

export const id = <T>(x: T) => x;

export const clearRemoteData = createAction('clearRemoteData');
export type RemoteDataSet = [Purchase, Product, Tag[]][];
export const setRemoteData = createAction<RemoteDataSet>('setRemoteData');
