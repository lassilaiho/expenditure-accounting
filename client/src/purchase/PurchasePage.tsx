import { Paper } from '@material-ui/core';
import Big from 'big.js';
import moment from 'moment';
import React from 'react';
import { Redirect, useParams } from 'react-router-dom';

import BackButton from '../common/BackButton';
import Scaffold from '../common/Scaffold';
import {
  getPurchaseById,
  useApi,
  useAppDispatch,
  useAppSelector,
} from '../data/store';
import CenteredLoader from '../common/CenteredLoader';
import PurchaseEditor from './PurchaseEditor';

interface PurchasePageParams {
  id: string;
}

const PurchasePage: React.FC = () => {
  const api = useApi();
  const dispatch = useAppDispatch();
  const { id: idParam } = useParams<PurchasePageParams>();
  const id = parseInt(idParam ?? '');
  let purchase = useAppSelector(getPurchaseById(id));

  if (idParam === 'new') {
    purchase = {
      id: -1,
      product: -1,
      date: moment.utc(),
      quantity: new Big(1),
      price: new Big(1),
      tags: [],
    };
    return (
      <PurchaseEditor
        purchase={purchase}
        onSave={update => dispatch(api.addPurchase(update))}
      />
    );
  }
  if (isNaN(id)) {
    return <Redirect to='/' />;
  }
  if (purchase) {
    return (
      <PurchaseEditor
        purchase={purchase}
        onSave={update => dispatch(api.updatePurchase(update))}
      />
    );
  }
  return (
    <Scaffold
      nav={<BackButton />}
      title='Edit Purchase'
      content={
        <Paper>
          <CenteredLoader />
        </Paper>
      }
    />
  );
};

export default PurchasePage;
