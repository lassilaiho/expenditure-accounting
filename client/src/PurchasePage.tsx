import { Paper } from '@material-ui/core';
import Big from 'big.js';
import { observer } from 'mobx-react';
import moment from 'moment';
import React from 'react';
import { Redirect, useParams } from 'react-router-dom';

import BackButton from './BackButton';
import Scaffold from './Scaffold';
import { Product, Purchase, useStore } from './data/store';
import CenteredLoader from './CenteredLoader';
import PurchaseEditor from './PurchaseEditor';

interface PurchasePageParams {
  id: string;
}

const PurchasePage: React.FC = observer(() => {
  const store = useStore();
  const { id: idParam } = useParams<PurchasePageParams>();
  if (idParam === 'new') {
    const purchase = new Purchase(
      -1,
      new Product(-1, ''),
      moment.utc(),
      new Big(1),
      new Big(1),
      [],
    );
    return <PurchaseEditor
      purchase={purchase}
      onSave={() => store.addPurchase(purchase)} />;
  }
  const id = parseInt(idParam ?? '');
  if (isNaN(id)) {
    return <Redirect to='/' />;
  }
  const purchase = store.purchasesById.get(id);
  if (purchase) {
    return <PurchaseEditor
      purchase={purchase}
      onSave={() => store.updatePurchase(purchase.id)} />;
  }
  return <Scaffold
    nav={<BackButton />}
    title='Edit Purchase'
    content={
      <Paper>
        <CenteredLoader />
      </Paper>
    }
  />;
});

export default PurchasePage;
