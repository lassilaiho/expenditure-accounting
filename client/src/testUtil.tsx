import MomentUtil from '@date-io/moment';
import { MuiPickersUtilsProvider } from '@material-ui/pickers';
import { render } from '@testing-library/react';
import React, { useRef } from 'react';
import { MemoryRouter } from 'react-router-dom';

import { fakeApi } from './data/fakeApi';
import { Store, StoreContext } from './data/store';

const TestProviders = (defaultStore?: Store, history?: string[]) => (
  props: any,
) => {
  const store = useRef<Store | null>(null);
  if (store.current === null) {
    if (defaultStore === undefined) {
      store.current = new Store(fakeApi);
    } else {
      store.current = defaultStore;
    }
  }
  return (
    <StoreContext.Provider value={store.current}>
      <MuiPickersUtilsProvider utils={MomentUtil}>
        <MemoryRouter initialEntries={history}>{props.children}</MemoryRouter>
      </MuiPickersUtilsProvider>
    </StoreContext.Provider>
  );
};

export interface CustomRenderOptions {
  store?: Store;
  history?: string[];
}

const customRender = (ui: any, options?: CustomRenderOptions) => {
  let store, history;
  if (options) {
    store = options.store;
    history = options.history;
  }
  return render(ui, { wrapper: TestProviders(store, history) });
};

export * from '@testing-library/react';

export { customRender as render };
