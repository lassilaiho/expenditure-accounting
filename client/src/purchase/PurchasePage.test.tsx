import { Store } from '@reduxjs/toolkit';
import moment from 'moment';
import React from 'react';
import { Route } from 'react-router-dom';
import { newStore, RootState } from '../data/store';

import { fakeApi, render as testUtilRender } from '../testUtil';
import PurchasePage from './PurchasePage';

const render = (id: string, store?: Store<RootState>) => {
  if (!store) {
    store = newStore();
  }
  return testUtilRender(
    <Route path='/purchases/:id'>
      <PurchasePage />
    </Route>,
    { store, history: ['/purchases/' + id] },
  );
};

test('renders correctly with invalid path', () => {
  const { asFragment } = render('asd');
  expect(asFragment()).toMatchSnapshot();
});

test('renders correctly with nonexistent purchase', async () => {
  const store = newStore();
  await store.dispatch(fakeApi.reloadData);
  const { asFragment } = render('42', store);
  expect(asFragment()).toMatchSnapshot();
});

test('renders correctly with existing purchase', async () => {
  const store = newStore();
  await store.dispatch(fakeApi.reloadData);
  const { asFragment } = render('3', store);
  expect(asFragment()).toMatchSnapshot();
});

test('renders correctly with new purchase', async () => {
  const original = moment.utc;
  moment.utc = () => original('2020-01-01');

  const store = newStore();
  await store.dispatch(fakeApi.reloadData);
  const { asFragment } = render('new', store);
  expect(asFragment()).toMatchSnapshot();

  moment.utc = original;
});
