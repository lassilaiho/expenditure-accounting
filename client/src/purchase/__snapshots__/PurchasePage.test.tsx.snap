// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with existing purchase 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-7"
  >
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <h6
          class="MuiTypography-root makeStyles-title-9 MuiTypography-h6"
        >
          Edit Purchase
        </h6>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M9 16.2L4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </header>
    <div
      class="makeStyles-contentRoot-8"
    >
      <div
        class="MuiContainer-root MuiContainer-fixed MuiContainer-maxWidthLg"
        style="width: 100%; height: 100%;"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            class="MuiBox-root MuiBox-root-10"
          >
            <div
              class="MuiFormGroup-root"
            >
              <div
                aria-expanded="false"
                class="MuiAutocomplete-root MuiAutocomplete-hasClearIcon"
                role="combobox"
              >
                <div
                  class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
                    data-shrink="true"
                    for="product-autocomplete"
                    id="product-autocomplete-label"
                  >
                    Product
                  </label>
                  <div
                    class="MuiInputBase-root MuiInput-root MuiInput-underline MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
                  >
                    <input
                      aria-autocomplete="list"
                      aria-invalid="false"
                      autocapitalize="none"
                      autocomplete="off"
                      class="MuiInputBase-input MuiInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd"
                      id="product-autocomplete"
                      spellcheck="false"
                      type="text"
                      value="product 2"
                    />
                    <div
                      class="MuiAutocomplete-endAdornment"
                    >
                      <button
                        aria-label="Clear"
                        class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator MuiAutocomplete-clearIndicatorDirty"
                        tabindex="-1"
                        title="Clear"
                        type="button"
                      >
                        <span
                          class="MuiIconButton-label"
                        >
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                            />
                          </svg>
                        </span>
                        <span
                          class="MuiTouchRipple-root"
                        />
                      </button>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
                  data-shrink="true"
                >
                  Purchase Date
                </label>
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
                    type="text"
                    value="14.03.2020"
                  />
                  <div
                    class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
                  >
                    <button
                      class="MuiButtonBase-root MuiIconButton-root"
                      tabindex="0"
                      type="button"
                    >
                      <span
                        class="MuiIconButton-label"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                          />
                          <path
                            d="M0 0h24v24H0z"
                            fill="none"
                          />
                        </svg>
                      </span>
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </button>
                  </div>
                </div>
              </div>
              <div
                class="MuiFormControl-root MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
                  data-shrink="true"
                >
                  Price
                </label>
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
                    type="number"
                    value="0.2"
                  />
                  <div
                    class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
                    >
                      â‚¬
                    </p>
                  </div>
                </div>
              </div>
              <div
                class="MuiFormControl-root MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
                  data-shrink="true"
                >
                  Quantity
                </label>
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiInput-input"
                    type="number"
                    value="0.8"
                  />
                </div>
              </div>
              <div
                aria-expanded="false"
                class="MuiAutocomplete-root MuiAutocomplete-hasClearIcon"
                role="combobox"
              >
                <div
                  class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink"
                    data-shrink="true"
                    for="tag-autocomplete"
                    id="tag-autocomplete-label"
                  >
                    Tags
                  </label>
                  <div
                    class="MuiInputBase-root MuiInput-root MuiInput-underline MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedStart MuiInputBase-adornedEnd"
                  >
                    <div
                      class="MuiChip-root MuiAutocomplete-tag MuiChip-deletable"
                      data-tag-index="0"
                      role="button"
                      tabindex="-1"
                    >
                      <span
                        class="MuiChip-label"
                      >
                        tag 1
                      </span>
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiChip-deleteIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
                        />
                      </svg>
                    </div>
                    <div
                      class="MuiChip-root MuiAutocomplete-tag MuiChip-deletable"
                      data-tag-index="1"
                      role="button"
                      tabindex="-1"
                    >
                      <span
                        class="MuiChip-label"
                      >
                        tag 2
                      </span>
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiChip-deleteIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
                        />
                      </svg>
                    </div>
                    <div
                      class="MuiChip-root MuiAutocomplete-tag MuiChip-deletable"
                      data-tag-index="2"
                      role="button"
                      tabindex="-1"
                    >
                      <span
                        class="MuiChip-label"
                      >
                        tag 3
                      </span>
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiChip-deleteIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
                        />
                      </svg>
                    </div>
                    <input
                      aria-autocomplete="list"
                      aria-invalid="false"
                      autocapitalize="none"
                      autocomplete="off"
                      class="MuiInputBase-input MuiInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd"
                      id="tag-autocomplete"
                      placeholder="Add Tag"
                      spellcheck="false"
                      type="text"
                      value=""
                    />
                    <div
                      class="MuiAutocomplete-endAdornment"
                    >
                      <button
                        aria-label="Clear"
                        class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator MuiAutocomplete-clearIndicatorDirty"
                        tabindex="-1"
                        title="Clear"
                        type="button"
                      >
                        <span
                          class="MuiIconButton-label"
                        >
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                            />
                          </svg>
                        </span>
                        <span
                          class="MuiTouchRipple-root"
                        />
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders correctly with invalid path 1`] = `<DocumentFragment />`;

exports[`renders correctly with new purchase 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-12"
  >
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <h6
          class="MuiTypography-root makeStyles-title-14 MuiTypography-h6"
        >
          Edit Purchase
        </h6>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M9 16.2L4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </header>
    <div
      class="makeStyles-contentRoot-13"
    >
      <div
        class="MuiContainer-root MuiContainer-fixed MuiContainer-maxWidthLg"
        style="width: 100%; height: 100%;"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            class="MuiBox-root MuiBox-root-15"
          >
            <div
              class="MuiFormGroup-root"
            >
              <div
                aria-expanded="false"
                class="MuiAutocomplete-root MuiAutocomplete-hasClearIcon"
                role="combobox"
              >
                <div
                  class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
                    data-shrink="false"
                    for="product-autocomplete"
                    id="product-autocomplete-label"
                  >
                    Product
                  </label>
                  <div
                    class="MuiInputBase-root MuiInput-root MuiInput-underline MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
                  >
                    <input
                      aria-autocomplete="list"
                      aria-invalid="false"
                      autocapitalize="none"
                      autocomplete="off"
                      class="MuiInputBase-input MuiInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd"
                      id="product-autocomplete"
                      spellcheck="false"
                      type="text"
                      value=""
                    />
                    <div
                      class="MuiAutocomplete-endAdornment"
                    >
                      <button
                        aria-label="Clear"
                        class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator MuiAutocomplete-clearIndicatorDirty"
                        tabindex="-1"
                        title="Clear"
                        type="button"
                      >
                        <span
                          class="MuiIconButton-label"
                        >
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                            />
                          </svg>
                        </span>
                        <span
                          class="MuiTouchRipple-root"
                        />
                      </button>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
                  data-shrink="true"
                >
                  Purchase Date
                </label>
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
                    type="text"
                    value="01.01.2020"
                  />
                  <div
                    class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
                  >
                    <button
                      class="MuiButtonBase-root MuiIconButton-root"
                      tabindex="0"
                      type="button"
                    >
                      <span
                        class="MuiIconButton-label"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                          />
                          <path
                            d="M0 0h24v24H0z"
                            fill="none"
                          />
                        </svg>
                      </span>
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </button>
                  </div>
                </div>
              </div>
              <div
                class="MuiFormControl-root MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
                  data-shrink="true"
                >
                  Price
                </label>
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
                    type="number"
                    value="1"
                  />
                  <div
                    class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
                    >
                      â‚¬
                    </p>
                  </div>
                </div>
              </div>
              <div
                class="MuiFormControl-root MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
                  data-shrink="true"
                >
                  Quantity
                </label>
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiInput-input"
                    type="number"
                    value="1"
                  />
                </div>
              </div>
              <div
                aria-expanded="false"
                class="MuiAutocomplete-root MuiAutocomplete-hasClearIcon"
                role="combobox"
              >
                <div
                  class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
                    data-shrink="false"
                    for="tag-autocomplete"
                    id="tag-autocomplete-label"
                  >
                    Tags
                  </label>
                  <div
                    class="MuiInputBase-root MuiInput-root MuiInput-underline MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
                  >
                    <input
                      aria-autocomplete="list"
                      aria-invalid="false"
                      autocapitalize="none"
                      autocomplete="off"
                      class="MuiInputBase-input MuiInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd"
                      id="tag-autocomplete"
                      placeholder="Add Tag"
                      spellcheck="false"
                      type="text"
                      value=""
                    />
                    <div
                      class="MuiAutocomplete-endAdornment"
                    >
                      <button
                        aria-label="Clear"
                        class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator"
                        tabindex="-1"
                        title="Clear"
                        type="button"
                      >
                        <span
                          class="MuiIconButton-label"
                        >
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                            />
                          </svg>
                        </span>
                        <span
                          class="MuiTouchRipple-root"
                        />
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders correctly with nonexistent purchase 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-2"
  >
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <h6
          class="MuiTypography-root makeStyles-title-4 MuiTypography-h6"
        >
          Edit Purchase
        </h6>
      </div>
    </header>
    <div
      class="makeStyles-contentRoot-3"
    >
      <div
        class="MuiContainer-root MuiContainer-fixed MuiContainer-maxWidthLg"
        style="width: 100%; height: 100%;"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            class="makeStyles-centeredContainer-5"
          >
            <div
              class="MuiCircularProgress-root MuiCircularProgress-colorSecondary MuiCircularProgress-indeterminate"
              role="progressbar"
              style="width: 40px; height: 40px;"
            >
              <svg
                class="MuiCircularProgress-svg"
                viewBox="22 22 44 44"
              >
                <circle
                  class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                  cx="44"
                  cy="44"
                  fill="none"
                  r="20.2"
                  stroke-width="3.6"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
